File: python/aoc2021/day21.py
import itertools
import collections
import functools

class GameState():
    def __init__(self, p1pos, p2pos, p1score, p2score, nextroll, nextplayer):
        self.p1pos = p1pos
        self.p2pos = p2pos
        self.p1score = p1score
        self.p2score = p2score
        self.nextroll = nextroll
        self.nextplayer = nextplayer

    def __repr__(self):
        return f'GameState=({self.p1pos} {self.p2pos} {self.p1score} {self.p2score} {self.nextroll} {self.nextplayer})'

    def __eq__(self, other):
        return self.p1pos == other.p1pos and self.p2pos == other.p2pos and self.p1score == other.p1score and self.p2score == other.p2score and self.nextroll == other.nextroll and self.nextplayer == other.nextplayer

    def __hash__(self):
        return hash((self.p1pos, self.p2pos, self.p1score,
                     self.p2score, self.nextroll, self.nextplayer))

def next_step(pos, score, points):
    pos = pos + points
    if pos > 10:
        pos = (pos - 1) % 10 + 1
    return pos, score + pos

@functools.cache
def move(gamestate):
    points = gamestate.nextroll

    p1pos = gamestate.p1pos
    p1score = gamestate.p1score
    p2pos = gamestate.p2pos
    p2score = gamestate.p2score

    if gamestate.nextplayer == 1:
        p1pos, p1score = next_step(p1pos, p1score, points)
    else:
        p2pos, p2score = next_step(p2pos, p2score, points)

    nextplayer = 1 if gamestate.nextplayer == 2 else 2
    return GameState(p1pos, p2pos, p1score, p2score, gamestate.nextroll, nextplayer)


def rolls():
    return itertools.product([1, 2, 3], repeat=3)


def play(p1start, p2start):
    metaverse = collections.OrderedDict()
    for universe in rolls():
        metaverse[GameState(p1start, p2start, 0, 0, sum(universe), 1)] = 1

    wins = [0, 0]
    counter = 0
    while len(metaverse) > 0:
        if counter > 100000:
            print(wins, len(metaverse))
            print(move.cache_info())
            counter = 0
        counter += 1

        game, game_count = metaverse.popitem()
        next_game = move(game)

        if next_game.p1score >= 21:
            wins[0] += game_count
            continue
        if next_game.p2score >= 21:
            wins[1] += game_count
            continue

        for universe in rolls():
            ng = GameState(next_game.p1pos, next_game.p2pos,
                                  next_game.p1score, next_game.p2score, sum(universe), next_game.nextplayer)
            if ng in metaverse:
                metaverse[ng] += game_count
            else:
                metaverse[ng] = game_count

    return wins


if __name__ == '__main__':
    res = play(4, 8)
    print(res)

File: govent/day21.go
package main

import (
	"fmt"
)

func rolls() [][3]int {
	return [][3]int{
		{1, 1, 1}, 
		{1, 1, 2}, 
		{1, 1, 3}, 
		{1, 2, 1}, 
		{1, 2, 2}, 
		{1, 2, 3}, 
		{1, 3, 1}, 
		{1, 3, 2}, 
		{1, 3, 3}, 
		{2, 1, 1}, 
		{2, 1, 2}, 
		{2, 1, 3}, 
		{2, 2, 1}, 
		{2, 2, 2}, 
		{2, 2, 3}, 
		{2, 3, 1}, 
		{2, 3, 2}, 
		{2, 3, 3}, 
		{3, 1, 1}, 
		{3, 1, 2}, 
		{3, 1, 3}, 
		{3, 2, 1}, 
		{3, 2, 2}, 
		{3, 2, 3}, 
		{3, 3, 1}, 
		{3, 3, 2}, 
		{3, 3, 3},
	}
}

type gameState struct{
	p1_pos int
	p2_pos int
	p1_score int
	p2_score int
	next_player int
	next_roll [3]int
}

func move(game gameState) gameState {
	points := 0
	for _, v := range game.next_roll {
		points += v
	}
	if game.next_player == 1 {
		game.p1_pos += points
		if game.p1_pos > 10 {
			game.p1_pos = (game.p1_pos - 1) % 10 + 1
		}
		game.p1_score += game.p1_pos
	} else {
		game.p2_pos += points
		if game.p2_pos > 10 {
			game.p2_pos = (game.p2_pos - 1) % 10 + 1
		}
		game.p2_score += game.p2_pos
	}
	if game.next_player == 1 {
		game.next_player = 2
	} else {
		game.next_player = 1
	}

	return game
}

func main() {
	// stack := []gameState{}
	universes := make(map[gameState]int)
	for _, roll := range rolls() {
		universes[gameState{8, 2, 0, 0, 1, roll}] = 1
	}
	wins := make(map[string]int)
	wins["p1"] = 0
	wins["p2"] = 0

	for len(universes) > 0 {
		fmt.Println(wins, len(universes))

		for universe, count := range universes {
			delete(universes, universe)
			
			next_universe := move(universe)
			
			if next_universe.p1_score >= 21 {
				wins["p1"] += count
				continue
			}			
			if next_universe.p2_score >= 21 {
				wins["p2"] += count
				continue
			}

			for _, roll := range rolls() {
				game := gameState{next_universe.p1_pos, next_universe.p2_pos, next_universe.p1_score, next_universe.p2_score, next_universe.next_player, roll}
				_, prs := universes[game]
				if prs {
					universes[game] += count
				} else {
					universes[game] = count
				}
			}
		}
	}
	fmt.Println(wins["p1"])
	fmt.Println(wins["p2"])
}
